feat(agent/agentcontainers): support agent name in customization

DanielleMaywood opened the pull request:
Relates to https://github.com/coder/internal/issues/732

This PR supports specifying a name that will be used for the devcontainer agent in the customizations section of the devcontainer.json configuration file. 

A follow up PR will introduce a fallback that names agents based on the folder name before falling back to the container name.

DanielleMaywood committed:
feat(agent/agentcontainers): support agent name in customization

This PR supports specifying a name that will be used for the
devcontainer agent in the customizations section of the
devcontainer.json configuration file.

DanielleMaywood committed:
chore: add extra test case

Technically this doesn't really test anything new but it is good to have
an explicit test for anyways.

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1173,6 +1174,14 @@ func (api *API) maybeInjectSubAgentIntoContainerLocked(ctx context.Context, dc c
+			// NOTE(DanielleMaywood):
+			// We only want to take an agent name specified in the very last customization layer.
+			// This restricts the ability for a feature to specify the agent name. We may revisit
+			// this in the future, but for now we want to restrict this behavior.
```
Thinking out loud but technically, if only a feature specifies customizations, won't it be the last one allowing it to change the name?

DanielleMaywood replied:
Ooooh that is a good point. I'll see if I can test this behavior.

DanielleMaywood replied:
Based on some little testing I'm thinking your suspicion is correct. I'm not entirely sure how we address this but I'll update the comment so it isn't factually incorrect.

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1173,6 +1174,14 @@ func (api *API) maybeInjectSubAgentIntoContainerLocked(ctx context.Context, dc c
+			// This restricts the ability for a feature to specify the agent name. We may revisit
+			// this in the future, but for now we want to restrict this behavior.
+			if len(coderCustomization) > 0 {
+				possibleAgentName = coderCustomization[len(coderCustomization)-1].Name
```
If this is gibberish/invalid, we will simply fail agent creation. It'd be best to validate it before considering it (and log + rely on fallbacks if bad).

mafredri reviewed (commented):
I think we need some input validation but approach seems good to me otherwise 👍🏻

DanielleMaywood committed:
chore: ignore invalid names

DanielleMaywood committed:
refactor: include configuration and merged configuration

We want to be able to pick customizations that are only specified in the
top-level customizations stanza. That is, we do not want to accidentally
pick something specified in a feature (such as the agent name).

This refactor picks the agent name from configuration instead of the
merged configuration to ensure a feature cannot influence it.

mafredri reviewed (approved):
Solid! 👍🏻

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -28,6 +28,7 @@ import (
 	"github.com/coder/coder/v2/coderd/httpapi"
 	"github.com/coder/coder/v2/codersdk"
 	"github.com/coder/coder/v2/codersdk/agentsdk"
+	"github.com/coder/coder/v2/provisioner"
```
Oh wait, does this import increase the size of the agent? I think it's better to copy over rather than depend on the import but if size increase is negligible then perhaps it's fine.

DanielleMaywood replied:
I've checked and they're it appears to be within the same MB before and after.
